apps/aefate/src/aefa_fate_op.erl

```diff
@@ -1247,6 +1261,134 @@ spend(Arg0, Arg1, ES0) ->
             aefa_fate:abort({value_does_not_match_type, Other, address}, ES2)
     end.
 
+create(Arg0, Arg1, Arg2, ES0) ->
+    ?AVAILABLE_FROM(?VM_FATE_SOPHIA_2, ES0),
+    {[?FATE_CONTRACT_BYTEARRAY(Code), InitArgsTypes, Value], ES1} =
+        get_op_args([Arg0, Arg1, Arg2], ES0),
+    create_common(Code, InitArgsTypes, Value, no_gas_cap, ES1).
+
+create_common(Code, InitArgsTypes, Value, GasCap, ES0) ->
+    {ContractPK, ES1} = create_contract({new, Code}, Value, ES0),
+    ReplaceInitResult
+        = fun(ES0_) ->
+                  %% init returns unit, so we get rid of it
+                  {{tuple, {}}, ES1_} = get_op_arg({stack, 0}, ES0_),
+                  write({stack, 0}, ?FATE_CONTRACT(ContractPK), ES1_)
+          end,
+    {ok, ES2} = remote_call_common(
+           ?FATE_CONTRACT(ContractPK),
+           init,
+           InitArgsTypes,
+           ?FATE_TYPEREP({tuple, []}),
+           0,
+           GasCap,
+           unprotected,
+           ReplaceInitResult,
+           ES1),
+    {jump, 0, ES2}.
+
+clone(Arg0, Arg1, Arg2, Arg3, ES0) ->
+    ?AVAILABLE_FROM(?VM_FATE_SOPHIA_2, ES0),
+    {[?FATE_CONTRACT(CloneePK), InitArgsTypes, Value, Prot], ES1} =
+        get_op_args([Arg0, Arg1, Arg2, Arg3], ES0),
+    clone_common(CloneePK, InitArgsTypes, Value, no_gas_cap, Prot, ES1).
+
+clone_g(Arg0, Arg1, Arg2, Arg3, Arg4, ES0) ->
+    ?AVAILABLE_FROM(?VM_FATE_SOPHIA_2, ES0),
+    {[?FATE_CONTRACT(CloneePK), InitArgsTypes, Value, GasCap, Prot], ES1} =
+        get_op_args([Arg0, Arg1, Arg2, Arg3, Arg4], ES0),
+    clone_common(CloneePK, InitArgsTypes, Value, {gas_cap, GasCap}, Prot, ES1).
+
+clone_common(CloneePK, InitArgsTypes, Value, GasCap, Prot, ES0) ->
+    Protected =
+        case Prot of
+            false -> unprotected;
+            true  -> protected;
+            _     -> aefa_fate:abort({value_does_not_match_type, Prot, bool}, ES0)
+        end,
+
+    {ok, FinalCloneePK} =
+        aefa_engine_state:contract_find_final_ref(CloneePK, ES0),
+
+    {ContractPK, ES1} = create_contract({clone, FinalCloneePK}, Value, ES0),
+    ReplaceInitResult
+        = fun(ES0_) ->
+                  %% init returns unit, so we get rid of it
+                  %% if the call was protected, rewrap the result
+                  case Protected of
+                      unprotected ->
+                          {{tuple, {}}, ES1_} = get_op_arg({stack, 0}, ES0_),
+                          write({stack, 0}, ?FATE_CONTRACT(ContractPK), ES1_);
+                      protected ->
+                          case get_op_arg({stack, 0}, ES0_) of
+                              {{variant,[0, 1], 1, {{tuple, {}}}}, ES1_} ->
+                                  write( {stack, 0}, make_some(?FATE_CONTRACT(ContractPK))
+                                       , ES1_
+                                       );
+                              {{variant, [0, 1], 0, {}}, ES1_} ->
```

Comment:
`aefa_engine_state:remove_contract` should be called here rather than below?! Or else the contract will remain if the init-function fails...

Reply:
Why? `{failed_protected_call, ES4} ->` case should trigger right on the `init` fail, won't it?

Reply:
I think the `{failed_protected_call, ES4}` will only hit if the checks for the remote call fail? This is why we have the continuation 🤔

Reply:
wait, it won't be called if the remote call aborts?!

Reply:
The protection seems to handle `abort` as well as all other errors or do I misunderstand something?
```
protect(unprotected, Action, _) -> Action();
protect(protected, Action, Recover) ->
    try Action() catch _:_ -> Recover() end.
```

Reply:
You also have the `check_return_type_protected` involved

Reply:
`remote_call_common` returns before jumping to the actual code (this is why you need the continuation!). The `{jump, 0, ES4}` in the `ok` case is what starts the execution of the `init` function. So clearly if the init call fails there's no way we could end up in `failed_protected_call` branch.

Reply:
done